### PR TITLE
Move on_epoch_end and epoch increment to after run_evaluation loop.

### PR DESCRIPTION
**Pull request by justinxzhao**
https://github.com/ludwig-ai/ludwig/pull/3668 inadvertently moved the `on_epoch_end` callback to be invoked after the last batch has been processed, but before evaluation.

While technically still valid, this is a change from the previous behavior of `on_epoch_end` being called after evaluation, which may come as a surprise to users assuming evaluation results being present in `on_epoch_end`.

Note: `on_eval_end` is the recommended callback for handling fresh eval metrics.

### CODE REVIEW

File: ludwig/trainers/trainer.py [hunk 1]
```diff
@@ -1192,6 +1185,13 @@ def _train_loop(
                 if should_break:
                     return should_break
```
Comment (jeffkinnison) [inline ludwig/trainers/trainer.py:1186]: Will we be one epoch off if we return here?

**Review by arnavgarg1 — APPROVED** [review]

File: ludwig/trainers/trainer.py (hunk 1 above)
Comment (jeffkinnison) [reply ludwig/trainers/trainer.py:1186]: I think when the end of an epoch lines up with eval and early stopping, the final epoch increment won't happen.

File: ludwig/trainers/trainer.py [hunk 2]
```diff
@@ -1189,8 +1183,26 @@ def _train_loop(
                     if self.is_coordinator():
                         progress_tracker.save(os.path.join(save_path, TRAINING_PROGRESS_TRACKER_FILE_NAME))
 
+                # If this was the last batch, then increment the epoch counter and invoke the `on_epoch_end` callback.
+                if not has_called_on_epoch_end and batcher.last_batch():
```
Comment (jeffkinnison) [inline ludwig/trainers/trainer.py:1187]: I'm not a huge fan of the duplicate code blocks. If we remove the early return below, this block should be unnecessary, right?

File: ludwig/trainers/trainer.py [hunk 3]
```diff
@@ -1189,8 +1183,26 @@ def _train_loop(
...
+                    progress_tracker.epoch += 1
+                    self.callback(lambda c: c.on_epoch_end(self, progress_tracker, save_path))
+
+                    # Ensures that we only call `on_epoch_end` once, for example, if evaluation is configured to run
+                    # several times per epoch.
+                    has_called_on_epoch_end = True
+
                 if should_break:
-                    return should_break
+                    return True
```
Comment (jeffkinnison) [inline ludwig/trainers/trainer.py:1196]: The early return here seems odd given that the only following piece of code is an epoch increment and callback that won't happen before the end of the epoch. What if we set `should_break = False` at the top of this method, then returned `should_break` at the end? That would let us remove the duplicate block above, remove this block, and make it so we don't need the `has_called_on_epoch_end` flag.

File: ludwig/trainers/trainer.py (hunk 2 above)
Comment (justinxzhao) [reply ludwig/trainers/trainer.py]: +1

File: ludwig/trainers/trainer.py (hunk 3 above)
Comment (justinxzhao) [reply ludwig/trainers/trainer.py]: Ah, that's a brain fart on my end -- thanks for the suggestion. Returning `should_break` is definitely simpler!

**Review by jeffkinnison — APPROVED** [review]
Nice! This looks great, and it simplifies the second half of `_train_loop` quite a bit

**Pull request merged**